README.md
# Homebridge Sunlight
[![npm-version](https://badgen.net/npm/v/homebridge-sunlight)](https://www.npmjs.com/package/homebridge-sunlight)
[![npm-downloads](https://badgen.net/npm/dt/homebridge-sunlight)](https://www.npmjs.com/package/homebridge-sunlight)


This is a plugin for [homebridge](https://github.com/nfarina/homebridge). It provides contact sensors based on sun position and clouds to automate sun protection. Sensors are opened when the sun is in a defined section of the sky (azimuth) and optionally if an [OpenWeather API key](https://openweathermap.org/api) is provided when the sky is not overcast and sun is above the horizon.


# Installation
Intall via hombridge GUI [homebridge-config-ui-x](https://github.com/oznu/homebridge-config-ui-x) or manually via:

1.  Install homebridge (if not already installed) using: `npm install -g homebridge`
2.  Install this plugin using: `npm install -g homebridge-sunlight`
3.  Update your configuration file (see below).

# Example Configuration

See `config-sample.json` for an example config. This plugin can also be configured through a GUI like [homebridge-config-ui-x](https://github.com/oznu/homebridge-config-ui-x).

## Platform Configuration

| Parameter  | Note                                                                  |
| ---------- | --------------------------------------------------------------------- |
| `lat`      | Latitude of the location the sun position should be calculated for   |
| `long`     | Longitude of the location the sun position should be calculated for  |
| `apikey`     | Your [OpenWeather API key](https://openweathermap.org/api), optional  |
| `sensors`  | Array of objects containing configuration for the sensors, see below |
| `debugLog` | Debug log output, optional, default: false                 |

## Sensors Configuration

Define contact sensors for one or more sections of the sky, e.g. for windows looking to different directions. Direction is specified in degrees of compass: 0° North, 90° East, 180° South, 270° West. To define overlaps you can use values between -360° and 0° as well as 360° and 720°.

| Parameter   | Note                                                                                                            |
| ----------- | --------------------------------------------------------------------------------------------------------------- |
| `name`           | Display name of the sensor                                                                                |
| `lowerThreshold` | Left side of sky section within which the sensor should activate |
| `upperThreshold` | Right side of sky section within which the sensor should activate |

**Thresholds example**: If you want the sensor to turn on when the sun is between 0° and 90° azimuth, set the lower threshold to 0 and the upper threshold to 90. See the example configuration file for a basic set-up (north, east, south, west).

For help or in case of issues please visit the [GitHub repository](https://github.com/Krillle/homebridge-sunlight/issues).    
This plugin is based on [homebridge-sunsensors](https://github.com/mfkrause/homebridge-sunsensors).


CHANGELOG.md
## Change Log

### v0.2.3 (2021/05/13 23:45)
- fix: device serial number must have at least 3 digits
- fix: avoid divergent sensor cloud state by checking cloud state at uneven times

### v0.2.2 (2021/04/13 01:00)
- feature: device model and manufacturer

### v0.2.1 (2021/04/12 23:45)
- fix: error handling
- fix: UI settings description text
- chore: GitHub description
- chore: Readme description text

### v0.2.0 (2021/04/12 14:00)
 Initial fully working version, based on fork from [homebridge-sunsensors](https://github.com/mfkrause/homebridge-sunsensors)


base/platform.js
const SunlightAccessory = require('./accessory');

let homebridge;

class SunlightPlatform {
  constructor(log, config) {
    this.config = config;
    this.log = log;
    this.accessories = [];

    // Initialize accessories
    this.sensors = {};
    config.sensors.forEach((sensorConfig) => {
      this.sensors[sensorConfig.name] = new SunlightAccessory(log, sensorConfig, config);
    });

    // Register new accessories after homebridge loaded
    homebridge.on('didFinishLaunching', this.registerAccessories.bind(this));
  }

  registerAccessories() {
    const { log, config } = this;

    // Unregister removed accessories first
    let tempAccessories = [];
    this.accessories.forEach((accessory) => {
      const configExists = config.sensors.find(
        (sensor) => UUIDGen.generate(sensor.name) === accessory.UUID,
      );

      if (!configExists) {
        log('Removing existing platform accessory from cache:', accessory.displayName);
        try {
          homebridge.unregisterPlatformAccessories('homebridge-sunlight', 'Sunlight', [accessory]);
        } catch (e) {
          log('Could not unregister platform accessory!', e);
        }
      } else {
        tempAccessories.push(accessory);
      }
    });
    this.accessories = tempAccessories;

    tempAccessories = [];
    // Update cached accessories
    if (this.accessories.length > 0) {
      this.accessories.forEach((accessory) => {
        log('Updating cached accesory:', accessory.displayName);
        const sensorConfig = config.sensors.find(
          (sensor) => sensor.name === accessory.displayName,
        );
        if (
          sensorConfig.lowerThreshold === undefined
          || sensorConfig.upperThreshold === undefined
          || typeof sensorConfig.lowerThreshold !== 'number'
          || typeof sensorConfig.upperThreshold !== 'number'
          || sensorConfig.lowerThreshold > 720
          || sensorConfig.lowerThreshold < -360
          || sensorConfig.upperThreshold > 720
          || sensorConfig.upperThreshold < -360) {
          log(`Error: Thresholds of sensor ${sensorConfig.name} are not correctly configured. Please refer to the README. Unregistering this cached accessory.`);
          try {
            homebridge.unregisterPlatformAccessories('homebridge-sunlight', 'Sunlight', [accessory]);
          } catch (e) {
            log('Could not unregister platform accessory!', e);
          }
        } else {
          const sensor = this.sensors[sensorConfig.name];
          sensor.setAccessory(accessory);
          tempAccessories.push(accessory);
        }

        // this.accessories[index] = this.sensors[accessory.displayName].initializeAccessory();
      });
      homebridge.updatePlatformAccessories('homebridge-sunlight', 'Sunlight', this.accessories);
    }
    const configuredAccessories = tempAccessories;
    this.accessories = [];

    // Initialize new accessoroies
    config.sensors.forEach((sensorConfig) => {
      const configured = configuredAccessories.find(
        (accessory) => accessory.UUID === UUIDGen.generate(sensorConfig.name),
      );
      if (configured) return;

      log('Registering accessory:', sensorConfig.name);

      if (
        sensorConfig.lowerThreshold === undefined
        || sensorConfig.upperThreshold === undefined
        || typeof sensorConfig.lowerThreshold !== 'number'
        || typeof sensorConfig.upperThreshold !== 'number'
        || sensorConfig.lowerThreshold > 720
        || sensorConfig.lowerThreshold < -360
        || sensorConfig.upperThreshold > 720
        || sensorConfig.upperThreshold < -360) {
        log(`Error: Thresholds of sensor ${sensorConfig.name} are not correctly configured. Please refer to the README.`);
        return;
      }

      const sensor = this.sensors[sensorConfig.name];
      if (!sensor.hasRegistered()) {
        this.accessories.push(sensor.initializeAccessory());
      }
    });

    // Collect all accessories after initialization to register them with homebridge
    if (this.accessories.length > 0) {
      homebridge.registerPlatformAccessories('homebridge-sunlight', 'Sunlight', this.accessories);
    }
  }

  configureAccessory(accessory) {
    this.accessories.push(accessory);
  }
}

/**
 * Set homebridge reference for platform, called from /index.js
 * @param {object} homebridgeRef The homebridge reference to use in the platform
 */
SunlightPlatform.setHomebridge = (homebridgeRef) => {
  homebridge = homebridgeRef;
};

module.exports = SunlightPlatform;


base/accessory.js
const suncalc = require('suncalc');
const request = require('request');

class SunlightAccessory {
  constructor(log, config, platformConfig) {
    this.accessory = null;
    this.registered = null;
    this.config = config;
    this.platformConfig = platformConfig;
    this.log = log;

    this.cachedWeatherObj = undefined;
    this.lastupdate = 0;
    if (this.platformConfig.apikey) {
      this.getWeather();
      setInterval(() => { this.getWeather(); }, 294997);
    }
  }

  getAccessory() {
    return this.accessory;
  }

  setAccessory(accessory) {
    this.accessory = accessory;
    this.setAccessoryEventHandlers();
  }

  hasRegistered() {
    return this.registered;
  }

  initializeAccessory() {
    const { config } = this;
    const { lowerThreshold, upperThreshold } = config;
    const uuid = UUIDGen.generate(config.name);
    const accessory = new Accessory(config.name, uuid);
    // Add Device Information
    accessory.getService(Service.AccessoryInformation)
      .setCharacteristic(Characteristic.Manufacturer, 'Krillle')
      .setCharacteristic(Characteristic.Model, 'Azimuth ' + lowerThreshold + '-' + upperThreshold)
      .setCharacteristic(Characteristic.SerialNumber, '---');

    const SensorService = accessory.addService(Service.ContactSensor, config.name);

    if (SensorService) {
      SensorService.getCharacteristic(Characteristic.ContactSensorState);
    }

    this.setAccessory(accessory);

    return accessory;
  }

  setRegistered(status) {
    this.registered = status;
  }

  setAccessoryEventHandlers() {
    const { log } = this;

    this.getAccessory().on('identify', (paired, callback) => {
      log(this.getAccessory().displayName, `Identify sensor, paired: ${paired}`);
      callback();
    });

    const SensorService = this.getAccessory().getService(Service.ContactSensor);

    if (SensorService) {
      SensorService
        .getCharacteristic(Characteristic.ContactSensorState)
        .on('get', this.getState.bind(this));

      SensorService.setCharacteristic(Characteristic.ContactSensorState, this.updateState());
      setInterval(() => {
        SensorService.setCharacteristic(Characteristic.ContactSensorState, this.updateState());
      }, 10007);
    }
  }

  updateState() {
    const { config, platformConfig, log } = this;
    const { lat, long, apikey } = platformConfig;
    const { lowerThreshold, upperThreshold } = config;
    const threshold = [lowerThreshold, upperThreshold];

    if (!lat || !long || typeof lat !== 'number' || typeof long !== 'number') {
      log('Error: Lat/Long incorrect. Please refer to the README.');
      return 0;
    }

    const sunPos = suncalc.getPosition(Date.now(), lat, long);
    let sunPosDegrees = Math.abs((sunPos.azimuth * 180) / Math.PI + 180);

    if (platformConfig.debugLog) log(`Current azimuth: ${sunPosDegrees}°`);

    if (threshold[0] > threshold[1]) {
      const tempThreshold = threshold[1];
      threshold[1] = threshold[0];
      threshold[0] = tempThreshold;
    }

    let newState;
    if (sunPosDegrees >= threshold[0] && sunPosDegrees <= threshold[1]) {
      newState = true;
    } else {
      newState = false;
    }
    if (threshold[0] < 0 && newState === false) {
      sunPosDegrees = -(360 - sunPosDegrees);
      if (sunPosDegrees >= threshold[0] && sunPosDegrees <= threshold[1]) {
        newState = true;
      } else {
        newState = false;
      }
    }
    if (threshold[1] > 360 && newState === false) {
      sunPosDegrees = 360 + sunPosDegrees;
      if (sunPosDegrees >= threshold[0] && sunPosDegrees <= threshold[1]) {
        newState = true;
      } else {
        newState = false;
      }
    }

    // Sun is in relevant azimuth range, lets check daylight and clouds
    if (newState && apikey) {
      let sunState = this.returnSunFromCache();
      let cloudState = this.returnCloudinessFromCache();
      if (platformConfig.debugLog) log(`Sun state: ${sunState}%, Cloud state: ${cloudState}%`);
      newState = sunState > 10 && sunState <90 && cloudState <= 25;
    }

    return newState;
  }

  getState(callback) {
    const { platformConfig, log } = this;
    const newState = this.updateState();

    callback(null, newState);
    if (platformConfig.debugLog) log(this.getAccessory().displayName, `getState: ${newState}`);
  }


  // - - - - - - - - Open Weather functions - - - - - - - -
  getWeather() {
    const { platformConfig, log } = this;
    const { lat, long, apikey } = platformConfig;

    // Only fetch new data once per minute
    if (!this.cachedWeatherObj || this.lastupdate + 60 < (new Date().getTime() / 1000 | 0)) {
      let p = new Promise((resolve, reject) => {
        var url = 'http://api.openweathermap.org/data/2.5/weather?appid=' + apikey + '&lat=' + lat + '&lon=' + long;
        if (platformConfig.debugLog) log("Checking weather: %s", url);
        request(url, function (error, response, responseBody) {
          if (error) {
              log("HTTP get weather function failed: %s", error.message);
              reject(error);
          } else {
              try {
                  if (platformConfig.debugLog) log("Server response:", responseBody);
                  this.cachedWeatherObj = JSON.parse(responseBody);
                  this.lastupdate = (new Date().getTime() / 1000);
                  log(`Sun state: ${this.returnSunFromCache()}%, Cloud state: ${this.returnCloudinessFromCache()}%`);
                  resolve(response.statusCode);
              } catch (error2) {
                  log("Getting Weather failed: %s", error2, responseBody);
                  reject(error2);
              }
          }
        }.bind(this))
      })
    }
  };

  returnCloudinessFromCache() {
    var value;
    if (this.cachedWeatherObj && this.cachedWeatherObj["clouds"]) {
        value = parseFloat(this.cachedWeatherObj["clouds"]["all"]);
    }
    return value;
  };

  returnSunFromCache() {
    var value;
    if (this.cachedWeatherObj && this.cachedWeatherObj["sys"]) {
        var sunrise = parseInt(this.cachedWeatherObj["sys"]["sunrise"]);
        var sunset = parseInt(this.cachedWeatherObj["sys"]["sunset"]);
        var now = Math.round(new Date().getTime() / 1000);
        if (now > sunset) {
            // It's already dark outside
            value = 100;
        } else if (now > sunrise) {
            // calculate how far though the day (where day is from sunrise to sunset) we are
            var intervalLen = (sunset - sunrise);
            value = (((now - sunrise) / intervalLen) * 100).toFixed(2);
        } else {
          value = 0;
        }
    }
    return value;
  };

}

module.exports = SunlightAccessory;
